module/state_synchronization/blob_io.go

```diff
@@ -0,0 +1,270 @@
+package state_synchronization
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/onflow/flow-go/network"
+)
+
+var ErrClosedBlobChannel = errors.New("send/receive on closed blob channel")
+
+type blobChannel struct {
+	blobs chan network.Blob
+	once  sync.Once
+	done  chan struct{}
+}
+
+func (bc *blobChannel) Send(blob network.Blob) error {
+	select {
+	case <-bc.done:
+		return ErrClosedBlobChannel
+	default:
+	}
+
+	select {
+	case bc.blobs <- blob:
+		return nil
+	case <-bc.done:
+		return ErrClosedBlobChannel
+	}
+}
+
+func (bc *blobChannel) Receive() (network.Blob, error) {
+	select {
+	case <-bc.done:
+		return nil, ErrClosedBlobChannel
+	default:
+	}
+
+	select {
+	case blob := <-bc.blobs:
+		return blob, nil
+	case <-bc.done:
+		return nil, ErrClosedBlobChannel
+	}
+}
+
+func (bc *blobChannel) Close() error {
+	bc.once.Do(func() { close(bc.done) })
+	return nil
+}
+
+type BlobChannelWriter struct {
+	maxBlobSize int
+	blobs       *blobChannel
+	buf         []byte
+	err         error
+}
+
+var _ io.WriteCloser = (*BlobChannelWriter)(nil)
+var _ io.ByteWriter = (*BlobChannelWriter)(nil)
+
+func (bw *BlobChannelWriter) Write(data []byte) (int, error) {
+	var n, m int
+
+	for n < len(data) && bw.err == nil {
+		m, bw.err = bw.write(data[n:])
+		n += m
```

Comment:
m is never negative because it represents the number of bytes written, which cannot be negative.

m can only be 0 if err != nil, in which case we would exit the loop.